v2/internal/reconcilers/arm/arm_client.go

```diff
@@ -0,0 +1,166 @@
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	client, ok := c.clients[key]
+	return client, ok
+}
+
+func (c armClients) GetClient(ctx context.Context, obj genruntime.ARMMetaObject, kubeClient kubeclient.Client, cloud cloud.Configuration) (*armClient, error) {
+	// Namespaced secret
+	secret, err := getSecret(ctx, kubeClient, obj.GetNamespace(), namespacedSecretName)
+	if err != nil {
+		return nil, err
+	}
+
+	if secret != nil {
+		nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
+		client, ok := c.Lookup(nsName.String())
+		// if client does not already exist, or secret data is changed, we need to get new token.
+		if !ok || !matchSecretData(client.secretData, secret.Data) {
+			credential, subscriptionID, err := c.newCredentialFromSecret(secret, nsName)
+			if err != nil {
+				return nil, err
+			}
+
+			newClient, err := genericarmclient.NewGenericClient(cloud, credential, subscriptionID, metrics.NewARMClientMetrics())
+			if err != nil {
+				return nil, err
+			}
+
+			armClient := newARMClient(newClient, secret.Data, nsName)
+			c.Register(armClient)
+			return armClient, nil
+		} else {
+			return client, nil
+		}
+	}
+	// If not found, default would be global
+	return c.globalClient, nil
+}
+
+func (c armClients) newCredentialFromSecret(secret *v1.Secret, nsName types.NamespacedName) (azcore.TokenCredential, string, error) {
+	subscriptionID, ok := secret.Data[config.SubscriptionIDVar]
+	if !ok {
+		return nil, "", resolver.NewSecretNotFoundError(nsName, errors.Errorf("Credential Secret %q does not contain key %q", nsName.String(), config.SubscriptionIDVar))
+	}
+	tenantID, ok := secret.Data[config.TenantIDVar]
+	if !ok {
+		return nil, "", resolver.NewSecretNotFoundError(nsName, errors.Errorf("Credential Secret %q does not contain key %q", nsName.String(), config.TenantIDVar))
+	}
+	clientID, ok := secret.Data[config.AzureClientID]
+	if !ok {
+		return nil, "", resolver.NewSecretNotFoundError(nsName, errors.Errorf("Credential Secret %q does not contain key %q", nsName.String(), config.AzureClientID))
+	}
+	clientSecret, ok := secret.Data[config.AzureClientSecret]
+	// TODO: We check this for now until we support Workload Identity. When we support workload identity for multitenancy, !ok would mean to check for workload identity.
+	if !ok {
+		return nil, "", resolver.NewSecretNotFoundError(nsName, errors.Errorf("Credential Secret %q does not contain key %q", nsName.String(), config.AzureClientSecret))
+	}
+
+	credential, err := azidentity.NewClientSecretCredential(string(tenantID), string(clientID), string(clientSecret), nil)
+	if err != nil {
+		return nil, "", errors.Wrap(err, errors.Errorf("Invalid Client Secret Credential for %q encountered", nsName.String()).Error())
+	}
+	return credential, string(subscriptionID), nil
+}
+
+func getSecret(ctx context.Context, kubeClient kubeclient.Client, namespace string, secretName string) (*v1.Secret, error) {
```

Comment:
Yeah and you could tack this on as a private method of `armClients` and if you took `kubeClient` as a parameter of `armClients` you wouldn't need to pass it here either.

Reply:
done!